File: Diya/Graphs/Clone_graph.java
class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) {
            return null;
        }

        // Map to track visited nodes and their clones
        HashMap<Node, Node> visited = new HashMap<>();

        // Perform DFS to clone the graph
        cloneNode(node, visited);

        // Return the clone of the original node
        return visited.get(node);
    }

    private void cloneNode(Node node, HashMap<Node, Node> visited) {
        // Create a clone for the current node
        Node clonedNode = new Node(node.val);
        visited.put(node, clonedNode);

        // Traverse neighbors
        for (Node neighbor : node.neighbors) {
            // If the neighbor is not visited, clone it recursively
            if (!visited.containsKey(neighbor)) {
                cloneNode(neighbor, visited);
            }
            // Add the cloned neighbor to the current cloned node
            visited.get(node).neighbors.add(visited.get(neighbor));
        }
    }
}

File: Karan/graphs/clone graph.py
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        clone_map = {}
        def dfs(original):
            if original in clone_map:
                return clone_map[original]
            cloned_node = Node(original.val)
            clone_map[original] = cloned_node
            for neighbor in original.neighbors:
                cloned_node.neighbors.append(dfs(neighbor))
            return cloned_node
        return dfs(node)

